### PR TITLE
builtin: add instr built-in function

### PR DESCRIPTION
**Pull request by cosmtrek**
Add `INSTR(str, substr)` built-in function

### CODE REVIEW

File: expression/builtin_string.go [hunk 1]
```diff
@@ -1565,7 +1565,29 @@ type builtinInstrSig struct {
 
 // See https://dev.mysql.com/doc/refman/5.6/en/string-functions.html#function_instr
 func (b *builtinInstrSig) eval(row []types.Datum) (d types.Datum, err error) {
-	return d, errFunctionNotExists.GenByArgs("instr")
+	args, err := b.evalArgs(row)
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+	// INSTR(str, substr)
+	if args[0].IsNull() {
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]:
``` go
if args[0].IsNull() || args[1].IsNull(){
    return
}
```

File: expression/builtin_string_test.go [hunk 2]
```diff
@@ -983,3 +983,33 @@ func (s *testEvaluatorSuite) TestLpad(c *C) {
...
+		Want interface{}
+	}{
+		{[]interface{}{"foobarbar", "bar"}, 4},
+		{[]interface{}{"xbar", "foobar"}, 0},
+
+		{[]interface{}{123456234, 234}, 2},
+		{[]interface{}{123456, 567}, 0},
+		{[]interface{}{1e10, 1e2}, 1},
+		{[]interface{}{1.234, ".234"}, 2},
+
```
Comment (zimulala) [inline expression/builtin_string_test.go]: Please add a test case like `SELECT INSTR('foobarbar', "'');`.

**Comment by XuHuaiyu** [comment]
@cosmtrek 
If possible, please check and register the builtin function you decide to implement in [issue](https://github.com/pingcap/tidb/issues/2863) to let others know, thank you!

**Comment by cosmtrek** [comment]
First thanks @spongedu 

I've just read the [INSTR()](https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_instr) carefully and realized another mistake except for the problem of ignoring utf8 specified by @spongedu 

> This function is multibyte safe, and is case sensitive only if at least one argument is a binary string.

We need to consider the type of arguments passed to `INSTR()` if binary or not. If two arguments are `string`, substring search should be **case-insensitive**, otherwise we performs case sensitive search.

Unfortunately, the implementation of `LOCATE()` function in master branch made this mistake as well...

File: expression/builtin_string.go [hunk 3]
```diff
@@ -1642,7 +1643,49 @@ type builtinInstrSig struct {
...
+	var substr string
+	if str, err = args[0].ToString(); err != nil {
+		return d, errors.Trace(err)
+	}
+	if substr, err = args[1].ToString(); err != nil {
+		return d, errors.Trace(err)
+	}
+
+	var caseSensitive bool
+	// INSTR performs case **insensitive** search by default
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: add `.` at the end of this comment.

File: expression/builtin_string.go [hunk 4]
```diff
@@ -1642,7 +1643,49 @@ type builtinInstrSig struct {
...
+	// While at least one argument is binary string we do case sensitive search
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: Ditto.

File: expression/builtin_string.go [hunk 5]
```diff
@@ -1642,7 +1643,49 @@ type builtinInstrSig struct {
...
+	if args[0].Kind() == types.KindBytes || args[1].Kind() == types.KindBytes {
+		caseSensitive = true
+	}
+
+	var pos int
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: var pos, idx int

File: expression/builtin_string.go [hunk 6]
```diff
@@ -1642,7 +1643,49 @@ type builtinInstrSig struct {
...
+	args, err := b.evalArgs(row)
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+	// INSTR(str, substr)
+	if args[0].IsNull() || args[1].IsNull() {
+		return d, nil
+	}
+
+	var str string
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: var str, substr string

**Comment by coocood** [comment]
LGTM
Thank you for your contribution.

**Pull request merged**